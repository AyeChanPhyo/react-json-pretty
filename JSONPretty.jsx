var React = require('react');

module.exports = React.createClass({
  // 格式化函数
  _replace: function (match, ind, key, val, tra) {
    var spanEnd = '</span>';
    var keySpan = '<span class=json-key>';
    var valSpan = '<span class=json-value>';
    var strSpan = '<span class=json-string>';
    var sps = ind || '';
    if (key) {
      sps = sps + '"' + keySpan + key.replace(/[": ]/g, '') + spanEnd + '": ';
    }
    if (val) {
      sps = sps + (val[0] == '"' ? strSpan : valSpan) + val + spanEnd;
    }
    return sps + (tra || '');
  },
  // JSON =》 HTML转换器
  _pretty: function (obj) {
    // 逐行匹配，列举：“key”: "value" | "key": value | "key": [ | "key": { | "key": [],| "Key": {},
    var regLine = /^( *)("[^"]+": )?("[^"]*"|[\w.+-]*)?([,[{]|\[\s*\],?|\{\s*\},?)?$/mg;
      return JSON.stringify(obj, null, 2)
        .replace(/&/g, '&amp;').replace(/\\"/g, '&quot;')
        .replace(/</g, '&lt;').replace(/>/g, '&gt;')
        .replace(regLine, this._replace);
  },
  render: function () {
    var json = this.props.json;

    if (typeof json === 'json') {
      json = JSON.parse(json);
    }

    return (
      <pre className='json-pretty' dangerouslySetInnerHTML={{__html: this._pretty(json)}}>
      </pre>
    );
  }
});
