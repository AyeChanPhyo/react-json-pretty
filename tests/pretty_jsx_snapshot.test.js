import React from 'react';
import JSONPretty from '../JSONPretty';
import renderer from 'react-test-renderer';

var obj2 = {
  aaaa: 1,
  bbbb: true,
  cccc: "ab\"cd\\",
  dddd: {
    eeee: 1,
    ffff: [
      {
        gggg: 3
      },
      2,
      "string"
    ]
  }
};

test('JSONPretty.jsx', () => {
	const json = renderer.create(
		<JSONPretty json={obj2}></JSONPretty>
  );

	let tree = json.toJSON();
	expect(tree).toMatchSnapshot();
});