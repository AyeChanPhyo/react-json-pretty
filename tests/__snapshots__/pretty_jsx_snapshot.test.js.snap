// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JSONPretty.jsx 1`] = `
<pre
  className="json-pretty"
  dangerouslySetInnerHTML={
    Object {
      "__html": "{
      \\"<span class=json-key>aaaa</span>\\": <span class=json-value>1</span>,
      \\"<span class=json-key>bbbb</span>\\": <span class=json-boolean>true</span>,
      \\"<span class=json-key>cccc</span>\\": <span class=json-string>\\"ab&quot;cd\\\\\\\\\\"</span>,
      \\"<span class=json-key>dddd</span>\\": {
        \\"<span class=json-key>eeee</span>\\": <span class=json-value>1</span>,
        \\"<span class=json-key>ffff</span>\\": [
          {
            \\"<span class=json-key>gggg</span>\\": <span class=json-value>3</span>
          },
          <span class=json-value>2</span>,
          <span class=json-string>\\"string\\"</span>
        ]
      }
    }",
    }
  }
/>
`;
